src/shogun/lib/any.h

```diff
@@ -0,0 +1,206 @@
+    class PointerValueAnyPolicy : public BaseAnyPolicy
+    {
+        public:
+            virtual void set(void** storage, const void* v) const
+            {
+                *(storage) = new T(*reinterpret_cast<T const*>(v));
+            }
+            virtual void clear(void** storage) const
+            {
+                delete reinterpret_cast<T*>(*storage);
+            }
+            virtual std::string type() const
+            {
+                return demangledType<T>();
+            }
+            virtual bool matches(const std::type_info& ti) const
+            {
+                return typeid(T) == ti;
+            }
+    };
+
+    /** @brief Class Any allows to store objects of arbitrary types. */
+    class Any
+    {
+    public:
+        struct Empty;
+
+        Any() : policy(selectPolicy<Empty>()), storage(nullptr)
+        {
+        }
+
+        template <typename T>
+        explicit Any(const T& v) : policy(selectPolicy<T>()), storage(nullptr)
+        {
+            policy->set(&storage, &v);
+        }
+
+        Any(const Any& other) : policy(other.policy), storage(nullptr)
+        {
+            policy->set(&storage, other.storage);
+        }
+
+        Any& operator=(const Any& other)
+        {
+            policy->clear(&storage);
+            policy = other.policy;
+            policy->set(&storage, other.storage);
+            return *(this);
+        }
+
+        ~Any()
+        {
+            policy->clear(&storage);
+        }
+
+        /** Casts hidden value to provided type, fails otherwise. */
+        template <typename T>
+        T& as() const
+        {
+            if (sameType<T>())
+            {
+                return *(reinterpret_cast<T*>(storage));
+            } 
+            else 
+            {
+                throw std::logic_error("Bad cast to" + demangledType<T>() + 
+                        " but the type is " + policy->type());
+            }
+        }
+
+        template <typename T>
+        inline bool sameType() const
+        {
+            return (policy == selectPolicy<T>()) || sameTypeFallback<T>();
+        }
+
+        template <typename T>
+        bool sameTypeFallback() const
+        {
+            return policy->matches(typeid(T));
+        }
+
+        bool empty() const
+        {
+            return sameType<Empty>();
+        }
+    private:
+        template <typename T>
+        static BaseAnyPolicy* selectPolicy()
+        {
+            typedef PointerValueAnyPolicy<T> Policy;
+            static Policy policy;
+            return &policy;
+        }
+    private:
+        BaseAnyPolicy* policy;
+        void* storage;
+    };
+
+    struct Any::Empty
+    {
+    };
+
+    namespace 
+    {
+        /** Erases value type i.e. converts it to Any */
+        template <typename T>
+        inline Any erase_type(const T& v)
+        {
+            return Any(v);
+        }
+
+        /** Tries to recall Any type, fails when type is wrong. */
```

Comment:
This code is hard to read, should be unit tested 100% coverage. Can you ensure that?